core/drivers/stm32_tamp.c

```diff
@@ -225,83 +496,159 @@ static TEE_Result stm32_tamp_parse_fdt(struct stm32_tamp_instance *tamp,
+		    !pdata->conf_data->priv_conf ||
+		    !pdata->conf_data->access_mask)
+			panic("Not enough memory capacity for TAMP RIF config");
+
+		for (i = 0; i < pdata->nb_rif_resources; i++)
+			stm32_rif_parse_cfg(fdt32_to_cpu(cuint[i]),
+					    pdata->conf_data, MAX_CID_SUPPORTED,
+					    TAMP_RIF_RESOURCES);
+	}
+
+	return TEE_SUCCESS;
 }
 
 static TEE_Result stm32_tamp_probe(const void *fdt, int node,
 				   const void *compat_data)
 {
-	struct stm32_tamp_instance *tamp = NULL;
 	uint32_t __maybe_unused revision = 0;
 	TEE_Result res = TEE_SUCCESS;
 	vaddr_t base = 0;
 
-	tamp = calloc(1, sizeof(*tamp));
-	if (!tamp)
+	stm32_tamp_dev = calloc(1, sizeof(*stm32_tamp_dev));
+	if (!stm32_tamp_dev)
 		return TEE_ERROR_OUT_OF_MEMORY;
 
-	res = stm32_tamp_parse_fdt(tamp, fdt, node, compat_data);
+	res = stm32_tamp_parse_fdt(fdt, node, compat_data);
 	if (res)
 		goto err;
 
-	clk_enable(tamp->clock);
+	if (clk_enable(stm32_tamp_dev->pdata.clock))
+		panic();
 
-	base = io_pa_or_va(&tamp->base, 1);
+	base = io_pa_or_va(&stm32_tamp_dev->pdata.base, 1);
 
-	tamp->hwconf1 = io_read32(base + _TAMP_HWCFGR1);
-	tamp->hwconf2 = io_read32(base + _TAMP_HWCFGR2);
+	stm32_tamp_dev->hwconf1 = io_read32(base + _TAMP_HWCFGR1);
+	stm32_tamp_dev->hwconf2 = io_read32(base + _TAMP_HWCFGR2);
 
 	revision = io_read32(base + _TAMP_VERR);
 	FMSG("STM32 TAMPER V%"PRIx32".%"PRIu32,
 	     (revision & _TAMP_VERR_MAJREV) >> 4, revision & _TAMP_VERR_MINREV);
 
-	if (!(tamp->hwconf2 & _TAMP_HWCFGR2_TZ)) {
+	if (!(stm32_tamp_dev->hwconf2 & _TAMP_HWCFGR2_TZ)) {
 		EMSG("TAMP doesn't support TrustZone");
 		res = TEE_ERROR_NOT_SUPPORTED;
 		goto err_clk;
 	}
 
-	/*
-	 * Enforce secure only access to protected TAMP registers.
-	 * Allow non-secure access to monotonic counter.
-	 */
-	stm32_tamp_set_secure(tamp, _TAMP_SECCFGR_TAMPSEC);
+	if (stm32_tamp_dev->pdata.compat->tags & TAMP_HAS_RIF_SUPPORT) {
+		apply_rif_config();
 
-	/*
-	 * Enforce privilege only access to TAMP registers, backup
-	 * registers and monotonic counter.
-	 */
-	stm32_tamp_set_privilege(tamp, _TAMP_PRIVCFG_TAMPPRIV |
-				 _TAMP_PRIVCFG_BKPRWPRIV |
-				 _TAMP_PRIVCFG_BKPWPRIV);
+		if (stm32_tamp_dev->pdata.is_tdcid) {
+			res = stm32_tamp_apply_bkpr_rif_conf();
+			if (res)
+				goto err;
+		}
+	} else {
+		/*
+		 * Enforce secure only access to protected TAMP registers.
+		 * Allow non-secure access to monotonic counter.
+		 */
+		stm32_tamp_set_secure(_TAMP_SECCFGR_TAMPSEC);
 
-	stm32_tamp_device = tamp;
+		/*
+		 * Enforce privilege only access to TAMP registers, backup
+		 * registers and monotonic counter.
+		 */
+		stm32_tamp_set_privilege(_TAMP_PRIVCFG_TAMPPRIV |
+					 _TAMP_PRIVCFG_BKPRWPRIV |
+					 _TAMP_PRIVCFG_BKPWPRIV);
+	}
 
 	return TEE_SUCCESS;
 
 err_clk:
-	clk_disable(tamp->clock);
+	clk_disable(stm32_tamp_dev->pdata.clock);
 err:
-	free(tamp);
+	if ((stm32_tamp_dev->pdata.compat->tags & TAMP_HAS_RIF_SUPPORT) &&
+	    stm32_tamp_dev->pdata.conf_data) {
```

Comment:
Testing only `stm32_tamp_dev->pdata.conf_data` is enough here.
Moreover, it will prevent a segfault if `stm32_tamp_parse_fdt()` return from line 497, before `pdata->compat` is set.

Ditto at lines 623/624.